pkg/cloud/services/ec2/instances.go

```diff
@@ -619,17 +619,10 @@ func (s *Service) runInstance(role string, i *infrav1.Instance) (*infrav1.Instan
 		resources := []string{ec2.ResourceTypeInstance, ec2.ResourceTypeVolume, ec2.ResourceTypeNetworkInterface}
 		for _, r := range resources {
 			spec := &ec2.TagSpecification{ResourceType: aws.String(r)}
-
-			// We need to sort keys for tests to work
-			keys := make([]string, 0, len(i.Tags))
-			for k := range i.Tags {
-				keys = append(keys, k)
-			}
-			sort.Strings(keys)
-			for _, key := range keys {
+			for tagKey, tagValue := range i.Tags {
 				spec.Tags = append(spec.Tags, &ec2.Tag{
-					Key:   aws.String(key),
-					Value: aws.String(i.Tags[key]),
```

Comment:
Isn't there a golintci check that would fail here because we're using `tagValue` instead of `i.Tags[tagKey]` with `i` being a pointer?

Reply:
Issues with using loop variables inside of the loop itself mostly become issues with things like goroutines, where the variable could change before it is used by the nested functions. In this case, we're only doing assignments in serial so there's no concern.

Reply:
Ahh - yeah makes sense. Thanks for the explanation!

Reply:
@AlexVulaj I think @iamkirkbater has a point here. What you said is true if you are using the _value_ of the loop variables. However `aws.String` is taking their address. Before Golang 1.22, loop variables are reused, so `spec.Tags` elements are pointing to the same value (the last one in `i.Tags`) at the end of the loop.

For example, the following go code:
```go
package main

import "fmt"

type Tag struct {
	key   *string
	value *string
}

func main() {
	tags := map[string]string{"key1": "value1", "key2": "value2"}
	saved := make([]*Tag, 0, 2)
	for key, val := range tags {
		saved = append(saved, &Tag{
			key:   &key,
			value: &val,
		})
	}
	fmt.Println(saved)
	for _, tag := range saved {
		fmt.Printf("%s: %s\n", *tag.key, *tag.value)
	}
}
```
outputs with Golang 1.21:
```
root@6b38e5f25656:/go/src# ./loop
[0xc0000140a0 0xc0000140b0]
key2: value2
key2: value2
```
with Golang 1.22:
```
$ ./loop
[0xc000096050 0xc000096080]
key2: value2
key1: value1
```

Reply:
I'll note - I'm surprised the tests didn't catch this, but it's probably because I'm running go 1.22

Reply:
I will say - I'm surprised that CI didn't catch this. Locally I'm running go 1.22 so it makes sense, but the green CI checks have me wondering if something's wrong with my matcher. I wasn't able to determine anything looking deeper locally.

Reply:
Ah, well we all learned something today! I'll revert that last change real quick then.

Reply:
For any other interested parties, this is the `aws.String` function in its entirety:
```
func String(v string) *string {
    return &v
}
```

So indeed as @r4f4 pointed out above, it's returning an address to the value passed in to the function, which is a different memory address than what's in the original caller.